lib/streams.gi

```diff
@@ -244,6 +243,104 @@ InstallOtherMethod( ReadAsFunction,
     [ IsInputStream ],
     READ_AS_FUNC_STREAM );
 
+#############################################################################
+##
+#F  ReadToList()  . . . . . . .  Reads commands and returns a list of results
+##
+##  <#GAPDoc Label="ReadToList">
+##  <ManSection>
+##  <Func Name="ReadToList" Arg='stream[,output_stream]'/>
+##
+##  <Description>
+##    Reads <A>stream</A> which is either an InputStream or a string containing
+##    complete GAP commands, executes them, and returns a list of results.
+##    The list of results is organized like the result of <C>READ_ALL_COMMANDS</C>.
+##    
+##    If the optional output stream <A>output_stream</A> argument is given, all
+##    additional output that happens while executing those commands is redirected
+##    into <A>output_stream</A>. This includes both standard and error output.
+##    If <A>output_stream</A> is <C>true</C>, the additional output is printed.
+##  </Description>
+##  </ManSection>
+##  <#/GAPDoc>
+##
+BindGlobal("ReadToList",
+function(stream, output_stream...)
+    local opened_stream, return_list, opened_dummy_output;
+
+    if IsString(stream) then
+        stream := InputTextString(stream);
+        opened_stream := true;
+    else
+        opened_stream := false;
+    fi;
+
+    opened_dummy_output := false;
+    if Length(output_stream) = 0 or output_stream[1] = false then
+        output_stream := OutputTextNone();
+        opened_dummy_output := true;
+    elif IsOutputStream(output_stream[1]) then
+        output_stream := output_stream[1];
+    elif output_stream[1] = true then
+        output_stream := 0;
+    else
+        Error("second argument must be an output stream or a boolean");
+    fi;
+
+    return_list := READ_ALL_COMMANDS(stream, output_stream, false);
+
+    if opened_stream then
+        CloseStream(stream);
+    fi;
+    if opened_dummy_output then
+        CloseStream(output_stream);
+    fi;
+
+    return return_list;
+end );
+
+#############################################################################
+##
+#F  ReadToString()  . . . . . . . . . . . Reads commands and returns a string
+##
+##  <#GAPDoc Label="ReadToString">
+##  <ManSection>
+##  <Func Name="ReadToString" Arg='stream[,output_stream]'/>
+##
+##  <Description>
+##    The same as <Ref Func="ReadToList"/>, except that it executes
+##    <Ref Oper="ViewObj"/> on every result and returns the resulting
+##    string.
+##  </Description>
+##  </ManSection>
+##  <#/GAPDoc>
+##
+BindGlobal("ReadToString",
+function(stream, output_stream...)
+    local return_list, return_string, return_stream, current_return;
+    
+    if Length(output_stream) > 0 then
+        output_stream := output_stream[1];
+    else
+        output_stream := false;
+    fi;
+
+    return_list := ReadToList(stream,output_stream);
+
+    return_string := "";
+    return_stream := OutputTextString(return_string,true);
+
+    OPEN_OUTPUT_STREAM(return_stream);
```

Comment:
I think the name `OPEN_OUTPUT_STREAM` is rather misleading or at least confusing -- after all, that stream already *was* opened. Perhaps `REDIRECT_OUTPUT_TO_STREAM`?

That said, I am somewhat unhappy about adding this back, after all, I just *removed* this functionality. I'd strongly prefer to hide this kernel implementation details from GAP code as much as possible. 

Moreover, I wonder if you really want this code to behave differently from the REPL, where execution and viewing are interleaved. A solution for both concerns would be to teach `READ_ALL_COMMANDS` to optionally invoke `ViewObj`, with output temporarily redirected (e.g. by adding yet another stream argument to `READ_ALL_COMMANDS`)

Reply:
Maybe we should stop mucking about with `READ_ALL_COMMANDS` and fix input streams to allow reading single commands from an input stream?

Reply:
For now, I will do as @fingolfin suggested and implement an output stream argument to READ_ALL_COMMANDS, I think. Lets see how this turns out.

Reply:
@markuspf sure, if you have a patch to do that, it'd be much better! But in the meantime, I think it's fairly easy to adjust `READ_ALL_COMMANDS` accordingly (or add a new kernel function doing this), so if the goal is to have something reasonable working quickly, I'd do that first, and switch to the "proper" solution once available.

That said, of course I use none of these things, so you guys can decide between yourselves how you want it :-)